File: ios-app/LEDMatrixController/LEDMatrixController/Services/Protocol.swift
//
//  Protocol.swift
//  LEDMatrixController
//
//  BLE Protocol constants matching the Raspberry Pi bridge service
//

import Foundation
import CoreBluetooth

/// BLE Protocol constants for LED Matrix communication
struct BLEProtocol {

    // MARK: - Service and Characteristic UUIDs

    static let serviceUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdef0")

    static let brightnessUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdef1")
    static let patternUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdef2")
    static let gameControlUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdef3")
    static let statusUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdef4")
    static let configUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdef5")
    static let powerLimitUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdef6")
    static let sleepScheduleUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdef7")
    static let frameStreamUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdef8")
    static let patternListUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdef9")
    static let gameListUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdefa")
    static let capabilitiesUUID = CBUUID(string: "12345678-1234-5678-1234-56789abcdeff")

    // MARK: - Pattern Names (indices 0-36)

    static let patterns: [String] = [
        // Solid Colors (0-3)
        "red", "green", "blue", "white",

        // Geometric (4-8)
        "corners", "cross", "checkerboard", "grid", "panels",

        // Animated Effects (9-13)
        "spiral", "wave", "fire", "plasma", "geometric_patterns",

        // Gradients & Colors (14-17)
        "rainbow", "color_gradients", "gradient_waves", "rgb_torch",

        // Natural Phenomena (18-26)
        "snow", "rain", "fireflies", "aquarium", "ocean_waves",
        "northern_lights", "starfield", "starry_night", "fireworks",

        // Complex Animations (27-32)
        "heart", "dna_helix", "kaleidoscope", "perlin_noise_flow",
        "matrix_rain", "lava_lamp",

        // Time-based (33-36)
        "elapsed", "sunset_sunrise", "sunset_sunrise_loop", "dot"
    ]

    // MARK: - Game Names (indices 0-4)

    static let games: [String] = [
        "snake",
        "pong",
        "tictactoe",
        "breakout",
        "tetris"
    ]

    // MARK: - Constants

    static let maxChunkSize = 500  // Maximum bytes per BLE write
    static let deviceName = "LED Matrix"  // BLE peripheral name to scan for
}

/// Game actions matching the bridge protocol
enum GameAction: UInt8 {
    case up = 0
    case down = 1
    case left = 2
    case right = 3
    case action = 4  // Select, fire, etc.
    case reset = 5
    case pause = 6
    case resume = 7
}

/// Pattern list response from device
struct PatternListResponse: Codable {
    let patterns: [String]
    let count: Int
}

/// Game list response from device
struct GameListResponse: Codable {
    let games: [String]
    let count: Int
}

/// Device capabilities response
struct DeviceCapabilities: Codable {
    let hasGames: Bool
    let hasPatterns: Bool
    let hasFrameStreaming: Bool
    let hasPowerLimiter: Bool
    let hasSleepScheduler: Bool
    let hasBrightnessControl: Bool
    let firmwareVersion: String

    enum CodingKeys: String, CodingKey {
        case hasGames = "has_games"
        case hasPatterns = "has_patterns"
        case hasFrameStreaming = "has_frame_streaming"
        case hasPowerLimiter = "has_power_limiter"
        case hasSleepScheduler = "has_sleep_scheduler"
        case hasBrightnessControl = "has_brightness_control"
        case firmwareVersion = "firmware_version"
    }
}

File: bluetooth_bridge/protocol.py
"""
BLE Protocol definitions for LED Matrix Controller

This module defines the Bluetooth Low Energy protocol used to communicate
between the iPhone app and the Raspberry Pi LED display.
"""

# BLE Service and Characteristic UUIDs
# Custom UUIDs generated for LED Matrix control
SERVICE_UUID = "12345678-1234-5678-1234-56789abcdef0"

CHAR_BRIGHTNESS_UUID = "12345678-1234-5678-1234-56789abcdef1"
CHAR_PATTERN_UUID = "12345678-1234-5678-1234-56789abcdef2"
CHAR_GAME_CONTROL_UUID = "12345678-1234-5678-1234-56789abcdef3"
CHAR_STATUS_UUID = "12345678-1234-5678-1234-56789abcdef4"
CHAR_CONFIG_UUID = "12345678-1234-5678-1234-56789abcdef5"
CHAR_POWER_LIMIT_UUID = "12345678-1234-5678-1234-56789abcdef6"
CHAR_SLEEP_SCHEDULE_UUID = "12345678-1234-5678-1234-56789abcdef7"
CHAR_FRAME_STREAM_UUID = "12345678-1234-5678-1234-56789abcdef8"
CHAR_PATTERN_LIST_UUID = "12345678-1234-5678-1234-56789abcdef9"

# Pattern indices (0-36 for 37 patterns)
PATTERNS = [
    # Solid Colors (0-3)
    "red", "green", "blue", "white",

    # Geometric (4-8)
    "corners", "cross", "checkerboard", "grid", "panels",

    # Animated Effects (9-13)
    "spiral", "wave", "fire", "plasma", "geometric_patterns",

    # Gradients & Colors (14-17)
    "rainbow", "color_gradients", "gradient_waves", "rgb_torch",

    # Natural Phenomena (18-26)
    "snow", "rain", "fireflies", "aquarium", "ocean_waves",
    "northern_lights", "starfield", "starry_night", "fireworks",

    # Complex Animations (27-32)
    "heart", "dna_helix", "kaleidoscope", "perlin_noise_flow",
    "matrix_rain", "lava_lamp",

    # Time-based (33-36)
    "elapsed", "sunset_sunrise", "sunset_sunrise_loop", "dot"
]

# Game indices (0-4)
GAMES = [
    "snake",
    "pong",
    "tictactoe",
    "breakout",
    "tetris"
]

# Game actions (0-7)
ACTIONS = [
    "up",
    "down",
    "left",
    "right",
    "action",
    "reset",
    "pause",
    "resume"
]

def get_pattern_name(index: int) -> str:
    """Get pattern name from index"""
    if 0 <= index < len(PATTERNS):
        return PATTERNS[index]
    return None

def get_pattern_index(name: str) -> int:
    """Get pattern index from name"""
    try:
        return PATTERNS.index(name)
    except ValueError:
        return -1

def get_game_name(index: int) -> str:
    """Get game name from index"""
    if 0 <= index < len(GAMES):
        return GAMES[index]
    return None

def get_action_name(index: int) -> str:
    """Get action name from index"""
    if 0 <= index < len(ACTIONS):
        return ACTIONS[index]
    return None

def get_pattern_list_json() -> str:
    """Get pattern list as JSON string"""
    import json
    return json.dumps({
        "patterns": PATTERNS,
        "count": len(PATTERNS)
    })

def get_game_list_json() -> str:
    """Get game list as JSON string"""
    import json
    return json.dumps({
        "games": GAMES,
        "count": len(GAMES)
    })

# Frame streaming constants
MAX_CHUNK_SIZE = 500  # Maximum bytes per BLE write
FRAME_TIMEOUT = 1.0   # Seconds before incomplete frame is discarded
